configure.q

```diff
@@ -2,6 +2,7 @@
 e:{-2 string x;exit 1}
 w:where{"B"$first@[system;string[x]," -c 'print(1)'";"0"]}@'pp:`python3`python                                                                      /python3 first
 py:{system string[$[count w;pp first w;e`python]]," -c \"",x,"\""}
+$["1"~first first py"import sysconfig as c;print(c.get_config_var('Py_ENABLE_SHARED'))";;'"shared not enabled"]
```

Comment:
There are many distributions that install statically linked Python, but still provide libpythonX.Y separately.  Examples include system Python on Ubuntu and the popular Anaconda distribution.  The most portable way to locate libpython on such systems is to call the `python-config` script.  For example, 

```bash
$ python-config --ldflags
-L/usr/local/opt/python36/Frameworks/Python.framework/Versions/3.6/lib/python3.6/config-3.6m-darwin -lpython3.6m -ldl -framework CoreFoundation
```

Reply:
@abalkin yes, you are right, `not Py_ENABLE_SHARED` doesn't mean there's no libpython..so.

do you think it's possible to locate the absolute path of libpython..so from python-config or sysconfig?  currently, configure.q `os.walk()`s some likely directories.  

on homebrew:
$ python-config --ldflags
-lpython3.5m -ldl -framework CoreFoundation -Wl,-stack_size,1000000 -framework CoreFoundation

meanwhile, we'll see what windows has to offer...

Reply:
Locating Python DLL is a pain, but python-config seems to get it right in most cases.  I wish I could say we do better in PyQ, but not really. Unfortunately,  python-config is sometimes installed as  python3-config or even  python3.6-config.

You can also take a look at how this problem is [solved](https://github.com/JuliaPy/PyCall.jl/blob/79909baa584a4ab6968e717c32eb130944d51f54/deps/build.jl#L61) in the  PyCall package for Julia.  A comment in their code says "TODO: other paths? python-config output? pyconfigvar("LDFLAGS")?"

Reply:
@abalkin hmmm - there's no python-config on win...
the only thing we do know for sure is that python3 or python is on the path.
if it's not on the path, we exit
from that point, we can find the path to the .so/.dll is using hints about directories from python

using V.v from sysconfig.get_config_var('py_version')
find the dylib:
see if there's locate on the system - if so, search for libpythonV.v[a-z]?.so\$ on mac and linux
if nothing found, use python's os.walk() searching for pythonVv.dll or libpythonV.v..so - using hints about directories from sysconfig
there don't seem to be hints about the location of the .so using python-config on homebrew

is exec_prefix reliably PYTHONHOME? (that's what's used in configure.q)

Reply:
What we do on Windows, is just [specify the DLL as "libpythonX.Y.dll"](https://github.com/KxSystems/pyq/blob/cbb5ba47eddb3d023436c23a54fa00ec6c1e4060/setup.py#L240) and let the runtime linker do the magic.

This also works for conda python on linux and system python on Ubuntu.

If python-config --ldflags does not print any -L option, it means libpython is on the system library path and a simple dlopen("libpythonX.Y.so") will probably succeed.

Another source of library paths that we use in pyq is the [ldconfig -p](https://github.com/KxSystems/pyq/blob/cbb5ba47eddb3d023436c23a54fa00ec6c1e4060/setup.py#L227) command.   We use that on Linux when we encounter a statically linked python.

Reply:
configure's main job is to locate DLLs that aren't accessible to the runtime linker.  In our experience, that includes anaconda where the only environment change is to extend PATH.

Reply:
I've just figured out what Anaconda does: it bakes the RPATH in the binaries compiled by distutils.   The relevant compile instructions can be found in `distutils.sysconfig.get_config_var('BLDSHARED')`.  On my Travis CI instance, I get

```python
 'BLDSHARED': 'gcc -pthread -shared -B '
              '/home/travis/miniconda/envs/test-env/compiler_compat '
              '-L/home/travis/miniconda/envs/test-env/lib '
              '-Wl,-rpath=/home/travis/miniconda/envs/test-env/lib '
              '-Wl,--no-as-needed -Wl,--sysroot=/',
```

Reply:
@abalkin when the dll is not found by the runtime linker, we add -rpath to link p.so.

-rpath is done by default now, as internal users had libpython3.5m.so in a custom place.

it would be nice to require users to add the directory to LD_LIBRARY_PATH, but i don't think that will happen

Reply:
@effbiae - did you see my previous comment?  It looks like our comments have crossed in flight.  As far as I can tell, on Linux you are passing an absolute path to `dlopen`, so it does not matter what the `RPATH` is set to at compilation time.  I think it is best to  call `dlopen` with just `v('LDLIBRARY')` and only specify the library path once in the linker `-rpath` option.

Reply:
@abalkin  nah - the linker loads libpython.so before the call to `dlopen`.
oleg and i spent some time dealing with this.  from my point of view, the best way is to not link with `-rpath` and have LD_LIBRARY_PATH/ldconfig give the right path to ld for libpythonV.v..so in the first place.  but users are going to think we're lazy if we don't find their dylib.

Reply:
@abalkin we probably will have to identify the kind of python installed... once we know it's anaconda, we look for BLDSHARED...  so we have a bunch of rules for each kind
 - anaconda
 - windows
 - mac homebrew
 - debian/ubuntu
 - ..
and we'll then have to provide some nice suggestions on error - like
 - mac homebrew: *install pyenv with PYTHON_CONFIGURE_OPTS="--enable-shared"*